pkg/controller/infra/infra_controller.go

```diff
@@ -118,99 +138,369 @@ type InfraController struct {
+	cLog := c.logger.WithField("job", key)
+	cLog.Debugln("started syncing job")
+	defer func() {
+		cLog.WithField("duration", time.Since(startTime)).Debugln("finished syncing job")
+	}()
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if len(ns) == 0 || len(name) == 0 {
+		return fmt.Errorf("invalid job key %q: either namespace or name is missing", key)
+	}
+
+	job, err := c.jobsLister.Jobs(ns).Get(name)
+	if errors.IsNotFound(err) {
+		cLog.Warnln("job not found")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	cluster, err := c.clusterForJob(job)
+	if err != nil {
+		return fmt.Errorf("could not retrieve cluster for job %s/%s: %v", job.Namespace, job.Name, err)
+	}
+	if cluster == nil {
+		return nil
+	}
+
+	// only proceed if this is the current provisioning job for the cluster
+	if cluster.Status.ProvisioningJob == nil || cluster.Status.ProvisioningJob.Name != job.Name {
+		return nil
+	}
+
+	return c.syncClusterStatusWithJob(cluster, job)
+}
+
+func (c *InfraController) syncClusterStatusWithJob(cluster *clusteroperator.Cluster, job *v1batch.Job) error {
+	cluster = cluster.DeepCopy()
+	now := metav1.Now()
+	jobCompleted := jobCondition(job, v1batch.JobComplete)
+	if jobCompleted != nil {
+		clusterProvisioning := clusterCondition(cluster, clusteroperator.ClusterHardwareProvisioning)
+		if clusterProvisioning != nil &&
+			clusterProvisioning.Status == kapi.ConditionTrue {
+			clusterProvisioning.Status = kapi.ConditionFalse
+			clusterProvisioning.LastTransitionTime = now
+			clusterProvisioning.LastProbeTime = now
+			clusterProvisioning.Reason = "JobCompleted"
+			clusterProvisioning.Message = fmt.Sprintf("Job %s/%s completed at %v", job.Namespace, job.Name, jobCompleted.LastTransitionTime)
+		}
+		clusterProvisioned := clusterCondition(cluster, clusteroperator.ClusterHardwareProvisioned)
+		if clusterProvisioned != nil &&
+			clusterProvisioned.Status == kapi.ConditionFalse {
+			clusterProvisioned.Status = kapi.ConditionTrue
+			clusterProvisioned.LastTransitionTime = now
+			clusterProvisioned.LastProbeTime = now
+			clusterProvisioned.Reason = "JobCompleted"
+			clusterProvisioning.Message = fmt.Sprintf("Job %s/%s completed at %v", job.Namespace, job.Name, jobCompleted.LastTransitionTime)
+			cluster.Status.Provisioned = true
+		}
+		if clusterProvisioned == nil {
+			cluster.Status.Conditions = append(cluster.Status.Conditions, clusteroperator.ClusterCondition{
+				Type:               clusteroperator.ClusterHardwareProvisioned,
+				Status:             kapi.ConditionTrue,
+				LastProbeTime:      now,
+				LastTransitionTime: now,
+				Reason:             "JobCompleted",
+				Message:            fmt.Sprintf("Job %s/%s completed at %v", job.Namespace, job.Name, jobCompleted.LastTransitionTime),
+			})
+			cluster.Status.Provisioned = true
+		}
+		cluster.Status.ProvisioningJob = nil
+	}
+	jobFailed := jobCondition(job, v1batch.JobFailed)
+	if jobFailed != nil {
```

Comment:
Is there value in a condition that states that the provisioning failed? As it stands, the only way to tell if the provisioning failed is to recognize that the provisioning condition is false (or not present) and the provisioned condition is not present. It may be difficult to distinguish this from a cluster that has not attempted provisioning (yet).